cores/esp8266/core_esp8266_noniso.cpp

```diff
@@ -40,11 +41,95 @@ char* ultoa(unsigned long value, char* result, int base) {
     return utoa((unsigned int)value, result, base);
 }
 
-char * dtostrf(double number, signed char width, unsigned char prec, char *s) {
-    char fmt[32];
-    sprintf(fmt, "%%%d.%df", width, prec);
-    sprintf(s, fmt, number);
+static char * __dtostrf(double number, signed char width, unsigned char prec, char *s) {
+    bool negative = false;
+
+    if (isnan(number)) {
+        strcpy(s, "nan");
+        return s;
+    }
+    if (isinf(number)) {
+        strcpy(s, "inf");
+        return s;
+    }
+
+    char* out = s;
+
+    int fillme = width; // how many cells to fill for the integer part
+    if (prec > 0) {
+        fillme -= (prec+1);
+    }
+
+    // Handle negative numbers
+    if (number < 0.0) {
+        negative = true;
+        fillme--;
+        number = -number;
+    }
+
+    // Round correctly so that print(1.999, 2) prints as "2.00"
+    // I optimized out most of the divisions
+    double rounding = 2.0;
+    for (uint8_t i = 0; i < prec; ++i)
+        rounding *= 10.0;
+    rounding = 1.0 / rounding;
+
+    number += rounding;
+
+    // Figure out how big our number really is
+    double tenpow = 1.0;
+    int digitcount = 1;
+    double nextpow;
+    while (number >= (nextpow = (10.0 * tenpow))) {
+        tenpow = nextpow;
+        digitcount++;
+    }
+
+    // minimal compensation for possible lack of precision (#7087 addition)
+    number *= 1 + std::numeric_limits<decltype(number)>::epsilon();
+
+    number /= tenpow;
+    fillme -= digitcount;
+
+    // Pad unused cells with spaces
+    while (fillme-- > 0) {
+        *out++ = ' ';
+    }
+
+    // Handle negative sign
+    if (negative) *out++ = '-';
+
+    // Print the digits, and if necessary, the decimal point
+    digitcount += prec;
+    int8_t digit = 0;
+    while (digitcount-- > 0) {
+        digit = (int8_t)number;
+        if (digit > 9) digit = 9; // insurance
+        *out++ = (char)('0' | digit);
+        if ((digitcount == prec) && (prec > 0)) {
+            *out++ = '.';
+        }
+        number -= digit;
+        number *= 10.0;
+    }
+
+    // make sure the string is terminated
+    *out = 0;
     return s;
 }
 
+// extern weak declaration (= function pointer) to check whether _printf_float is defined
+extern int _printf_float () __attribute__((__weak__));
+
+char * dtostrf(double number, signed char width, unsigned char prec, char *s) {
+    if (!_printf_float) {
```

Comment:
This function is used directly only by users/maintainers of builds that have float support disabled. Supporting sprintf in here doesn't make sense to me, it should be the __dtostrf body directly. That will give consistent results to the users regardless of whether float support is enabled or not.

Reply:
Whatever the case, please prefer snprintf over sprintf.

Reply:
> Whatever the case, please prefer snprintf over sprintf

Normally, I'd agree, but in this case it is a false sense of added security.  This API does not have a way of passing in sizeof(destbuff), and anything you artificially create in the function itself is bogus.

The "width" param is the integer width, not including the "." and the prec after the decimal point.  The "prec" param is the number of fractional digits.  No destination size can be calculated from these numbers...